### PR TITLE
Adding support for PBS cluster software

### PR DESCRIPTION
**Pull request by dranasinghe**
Add PBS  job scheduler. I do not have access to a server running PBS. I help Jeff to set up ARC on Dow cluster and modifications I did merge to the master. Modifications were tested on Dow cluster. I also added TXE1 submission scripts.

### CODE REVIEW

File: arc/job/ssh.py [hunk 1]
```diff
@@ -525,9 +533,14 @@ def check_job_status_in_stdout(job_id: int,
...
-    if status.lower() in ['r', 'qw', 't']:
-        return 'running'
+    if servers[server]['cluster_soft'].lower() == 'slurm':
+        status = status_line.split()[4]
+        if status.lower() in ['r', 'qw', 't']:
+            return 'running'
+    elif servers[server]['cluster_soft'].lower() == 'pbs':
+        status = status_line.split()[-2]
+        if status.lower() in ['r', 'q', 'c']:
+            return 'running'
```
Comment (alongd) [inline arc/job/ssh.py:543]: This function requires some more attention and organization.
In its current state, the last `else` will only trigger if the software is not Slurm or PBS, but then under it there's a line `elif servers[server]['cluster_soft'].lower() == 'slurm':` which now doesn't make sense.

File: arc/job/job.py [hunk 2]
```diff
@@ -1422,6 +1422,10 @@ def _get_additional_job_info(self):
                             lines1 = f.readlines()
                     content += ''.join([line for line in lines1])
                     content += '\n'
+        else:
```
Comment (alongd) [inline arc/job/job.py:1425]: Please change the commit message to something more informative, e.g.: `Added PBS cluster software to job`

File: arc/job/submitTest.py [hunk 3]
```diff
@@ -18,7 +18,7 @@ class TestSubmit(unittest.TestCase):
     def test_servers(self):
         """Test server keys in submit_scripts"""
         for server in submit_scripts.keys():
-            self.assertTrue(server in ['pharos', 'c3ddb', 'rmg'])
+            self.assertTrue(server in ['pharos', 'c3ddb', 'rmg', 'txe1'])
```
Comment (alongd) [inline arc/job/submitTest.py:21]: Please add this line as a separate commit: `Tests: Modification to submit test to account for TXE!`

File: arc/job/ssh.py (hunk 1 above)
Comment (alongd) [reply arc/job/ssh.py:543]: then, this chunk should be squashed into the job commit

File: arc/settings.py [hunk 4]
```diff
@@ -88,22 +88,28 @@
 }
 
 check_status_command = {'OGE': 'export SGE_ROOT=/opt/sge; /opt/sge/bin/lx24-amd64/qstat',
-                        'Slurm': '/usr/bin/squeue'}
```
Comment (alongd) [inline arc/settings.py:91]: Please rename the commit to `Added PBS cluster software commands to settings.py`

File: arc/settings.py [hunk 5]
```diff
@@ -88,22 +88,28 @@
...
 
 list_available_nodes_command = {'OGE': 'export SGE_ROOT=/opt/sge; /opt/sge/bin/lx24-amd64/qstat -f | grep "/8 " | grep "long" | grep -v "8/8"| grep -v "aAu"',
-                                'Slurm': 'sinfo -o "%n %t %O %E"'}
+                                'Slurm': 'sinfo -o "%n %t %O %E"',
+                                'PBS': 'pbsnodes'}
 
 submit_filename = {'OGE': 'submit.sh',
-                   'Slurm': 'submit.sl'}
+                   'Slurm': 'submit.sl',
+                   'PBS': 'PBS.sh'}
```
Comment (alongd) [inline arc/settings.py:108]: Can we make this `submit.sh` instead?

File: arc/job/ssh.py (hunk 1 above)
Comment (dranasinghe) [reply arc/job/ssh.py:543]: I rewrote the segment.

**Comment by alongd** [comment]
@dranasinghe, the tests are upset due to an error that seems trivial to fix:
```
======================================================================
FAIL: Test server keys in submit_scripts
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/ReactionMechanismGenerator/ARC/arc/job/submitTest.py", line 21, in test_servers
    self.assertTrue(server in ['pharos', 'c3ddb', 'rmg', 'txe1'])
AssertionError: False is not true
```

Can you take a look?

**Comment by dranasinghe** [comment]
Thanks, I will take a look
Duminda

File: arc/settings.py (hunk 4 above)
Comment (dranasinghe) [reply arc/settings.py:91]: Sorry saw this later. I will make the changes

File: arc/job/job.py [hunk 6]
```diff
@@ -1422,6 +1422,36 @@ def _get_additional_job_info(self):
                             lines1 = f.readlines()
                     content += ''.join([line for line in lines1])
                     content += '\n'
+        elif cluster_soft == 'pbs':
+            if self.server != 'local':
+                with SSHClient(self.server) as ssh:
```
[the remark this answers is not in the document]
Comment (dranasinghe) [reply arc/job/job.py:1427]: I made some changes. I add out.txt and err.txt to submission scripts with -o -e flags. So we can download the same files as stderr and stdout. This also makes the module much shorter.

File: arc/job/job.py [hunk 7]
```diff
@@ -1396,32 +1400,10 @@ def _get_additional_job_info(self):
             content += ''.join([line for line in lines1])
             content += '\n'
             content += ''.join([line for line in lines2])
-        elif cluster_soft == 'slurm':
+        else:
             if self.server != 'local':
-                with SSHClient(self.server) as ssh:
-                    response = ssh.list_dir(remote_path=self.remote_path)
-            else:
-                response = execute_command('ls -alF {0}'.format(self.local_path))
```
Comment (alongd) [inline arc/job/job.py:1404]: @dranasinghe, this line, for example, was removed but wasn't replaced (it's now missing in the new version). Is this on purpose?

Comment (dranasinghe) [reply arc/job/job.py:1404]: yes. I changed submit.py to have stderr and stdout have err.txt and out.txt. So we don't have to list the content and download the contents, we can directly download out.txt and err.txt.

**Comment by alongd** [comment]
@dranasinghe , can you rebase?

**Comment by dranasinghe** [comment]
Sure thing.

**Comment by alongd** [comment]
Thanks!!

**Pull request merged**